src/vos/vos_pool.c

```diff
@@ -1537,51 +1605,89 @@ lock_pool_memory(struct vos_pool *pool)
 		pool->vp_umm.umm_base);
 }
 
-/*
- * If successful, this function consumes ph, and closes it upon any error.
- * So the caller shall not close ph in any case.
- */
 static int
-pool_open(void *ph, struct vos_pool_df *pool_df, unsigned int flags, void *metrics,
-	  daos_handle_t *poh)
+pool_open_prep(uuid_t uuid, unsigned int flags, struct vos_pool **p_pool)
 {
 	struct vos_pool		*pool = NULL;
-	struct umem_attr	*uma;
 	struct d_uuid		 ukey;
 	int			 rc;
 
 	/* Create a new handle during open */
-	rc = pool_alloc(pool_df->pd_id, &pool); /* returned with refcount=1 */
+	rc = pool_alloc(uuid, &pool); /* returned with refcount=1 */
 	if (rc != 0) {
 		D_ERROR("Error allocating pool handle\n");
-		vos_pmemobj_close(ph);
 		return rc;
 	}
 
+	pool->vp_opening = 1;
+	pool->vp_sysdb = !!(flags & VOS_POF_SYSDB);
+	pool->vp_excl = !!(flags & VOS_POF_EXCL);
+	pool->vp_small = !!(flags & VOS_POF_SMALL);
+	pool->vp_rdb = !!(flags & VOS_POF_RDB);
+
+	/*
+	 * Insert the pool into the uuid hash table before full opened, because subsequent
+	 * pool_open process maybe yield and other pool_lookup for the same UUID will find
+	 * the in-processing pool_open, then wait there and avoid concurrent pool_open.
+	 */
+	uuid_copy(ukey.uuid, uuid);
+	rc = pool_link(pool, &ukey);
+	if (rc != 0)
+		pool_free(pool);
+	else
+		/* Here, two references on the pool: one is for myself, the other for the hash. */
+		*p_pool = pool;
+
+	return rc;
+}
+
+static int
+pool_open_post(struct umem_pool **p_ph, struct vos_pool_df *pool_df, unsigned int flags,
+	       void *metrics, struct vos_pool *pool, int ret)
+{
+	struct umem_attr	*uma;
+	int			 rc;
+
+	if (pool_df != 0) {
+		if (pool_df->pd_version >= VOS_POOL_DF_2_4)
+			pool->vp_feats |= VOS_POOL_FEAT_2_4;
+		if (pool_df->pd_version >= VOS_POOL_DF_2_6)
+			pool->vp_feats |= VOS_POOL_FEAT_2_6;
+		if (pool_df->pd_version >= VOS_POOL_DF_2_8)
+			pool->vp_feats |= VOS_POOL_FEAT_2_8;
+		pool->vp_pool_df = pool_df;
+	}
```

Comment:
Move this after following "ret != 0" checking?

Reply:
That is not matter since the pool will be unlinked from hash table if current open failed.